Source/NSubstitute/Core/GetCallSpec.cs

```diff
@@ -1,26 +1,45 @@
-﻿namespace NSubstitute.Core
+﻿using System;
+
+namespace NSubstitute.Core
 {
     public class GetCallSpec : IGetCallSpec
     {
-        private readonly ICallStack _callStack;
+        private readonly ICallCollection _callCollection;
         private readonly IPendingSpecification _pendingSpecification;
         private readonly ICallSpecificationFactory _callSpecificationFactory;
         private readonly ICallActions _callActions;
 
-        public GetCallSpec(ICallStack callStack, IPendingSpecification pendingSpecification,
+        public GetCallSpec(ICallCollection callCollection, IPendingSpecification pendingSpecification,
             ICallSpecificationFactory callSpecificationFactory, ICallActions callActions)
         {
-            _callStack = callStack;
+            _callCollection = callCollection;
             _pendingSpecification = pendingSpecification;
             _callSpecificationFactory = callSpecificationFactory;
             _callActions = callActions;
         }
 
-        public ICallSpecification FromLastCall(MatchArgs matchArgs)
+        public ICallSpecification FromPendingSpecification(MatchArgs matchArgs)
         {
-            return _pendingSpecification.HasPendingCallSpec()
-                               ? FromExistingSpec(_pendingSpecification.UseCallSpec(), matchArgs)
-                               : FromCall(_callStack.Pop(), matchArgs);
+            if (!_pendingSpecification.HasPendingCallSpecInfo())
+            {
+                throw new InvalidOperationException("No pending specification or previous call info.");
```

Comment:
Yeah, that method is a bit difficult to read. The logic stays the same. That's the case when type system should help with types like `Either`